sodym/named_dim_array_helper.py

```diff
@@ -0,0 +1,55 @@
+from einops import rearrange
+import numpy as np
+
+from .named_dim_arrays import Flow, NamedDimArray
+from .dimensions import Dimension, DimensionSet
+
+
+def ndarray_stack(ndarrays: list[NamedDimArray], dimension: Dimension) -> NamedDimArray:
+    """Stack a list of NamedDimArray objects using a new dimension.
+    Like numpy.stack with axis=-1, but for `NamedDimArray`s.
+    Method can be applied to `NamedDimArray`s, `StockArray`s, `Parameter`s and `Flow`s.
+    """
+    ndarray0 = ndarrays[0]
+    for ndarray in ndarrays[1:]:
+        if ndarray.dims != ndarray0.dims:
+            raise ValueError(
+                'Existing dimensions must be identical in all objects to be stacked.'
+            )
+    axis = len(ndarray0.dims.dimensions)
+    extended_values = np.stack([ndarray.values for ndarray in ndarrays], axis=axis)
+    extended_dimensions = DimensionSet(dimensions=ndarray0.dims.dimensions+[dimension])
+    if isinstance(ndarray0, Flow):
+        return Flow(
+            values=extended_values,
+            dims=extended_dimensions,
+            from_process=ndarray0.from_process,
+            to_process=ndarray0.to_process,
+            )
+    # otherwise, for NamedDimArray, StockArray or Parameter
+    return ndarray0.__class__(values=extended_values, dims=extended_dimensions)
```

Comment:
All of this could (and should, imo) be done via the NamedDimArray slice functionality. Its designed for this prupose, and orders dimensions and asserts correct shape. The current implementation is redundant to that. Also, I can't think of a case where returning a Flow from many flows makes sense. In this special case, you can still initialize it outside of the function and pass the values of the returned NamedDimArray to it (e.g. in the stock_stack function). The idea in the rest of the code was so far to use vanilla NamedDimArrays for temporary variables. 
Suggestion: 
```
    ndarray0 = ndarrays[0]
    if not len(ndarrays) == dimension.len
        raise ValueError('number of stacked arrays must be equal to length of new dimension')
    extended_dimensions = DimensionSet(dimensions=ndarray0.dims.dimensions+[dimension])
    nda_out = NamedDimArray(dims=extended_dimensions)
    for item, ndarray in zip(dimension.items, ndarrays): 
        nda_out[{dimension.name: item}] = ndarray
    return nda_out 
```
If you like to keep the flow, I'd be okay with 
```
    if isinstance(ndarray0, Flow): 
        nda_out = Flow(
            dims=extended_dimensions, 
            from_process=ndarray0.from_process,
            to_process=ndarray0.to_process,)
    else: 
        nda_out = ndarray0.__class__(dims=extended_dimensions)
```
Though this creates errors if a new subclass of NamedDimArray with additional attributes is created, so I prefer the first version.

Reply:
I tried this, but running into ValidationErrors, and failing tests (good that I wrote them). I haven't really worked out how the SubArrayHandler works and what it is used for... Guess I should learn this!

In general I think it makes sense to have this functionality available for Flows as well, and to keep the instance type. The usecases I see, are e.g. applying special case treatment (hacks) to a certain region / material / whatever. Then you can split, make your hacks, and then stack again to continue with the mfa equations.

Reply:
@JakobBD How do you feel about me editing the SubArrayHandler, so that (a) it doesn't remove all length-1 dimensions, only the one that we want to take a slice of, and (b) so that to_nda returns an instance of the original class?

Reply:
a) seems like fixing my bug, so yes, please, thanks!
b) I see two issues with that, but both of them can be solved: 
- I don't like the specific exception for the Flow class, because you'd have to adapt it if you were to come up with your own subclass of NamedDimArray. But you could add a method to NamedDimArray which gets the set difference of `self.__class__.__fields__` and `NamedDimArray.__fields__`, runs a `getattribute` on each of them, and returns the results as a `kwargs` dict.
- right now, any operation on one or several objects of NamedDimArray subclasses consistently return NamedDimArray objects, which is as a rule simple enough to make the behaviour predictable. If you deviate from that, let's find a new consistent way to determine the return type. This applies, for example, to the methods `cast_to`, `sum_nda_to`, `sum_nda_over`, and all the mathematical operations. I don't know if the list is complete.

Reply:
Not sure what to do with `cast_to`, `sum_nda_to` and `sum_nda_over`, maybe I need to see some usecases, but I guess it also makes sense to change this to preserve type too.